### PR TITLE
Add support to install extensions

### PR DESCRIPTION
**Pull request by jsuchome**
See https://github.com/fuseml/fuseml/issues/161

First commit contains the blueprint.

TODO (not necessary in this PR)

- [x] install extension: helm type
- [x] uninstall extension: helm type
- [x] option to pass extension registry path
- [x] install extension: script type
- [x] uninstall extension: script type
- [x] install extension: manifest type
- [x] uninstall extension: manifest type
- [ ] separate extension command
- [ ] extension registry for FuseML ? (so we can list extensions)
- [x] solve relative paths (helm chart, values.yaml, scripts...)
- [ ] wait for pods to exist/be ready after each installation step

### CODE REVIEW

File: docs/blueprints/001-installation-of-extensions.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+  `fuseml-installer extension`
+
+  This should have several sub commands, like
+
+  * `fuseml-installer extension list` - list installed extensions
+  * `fuseml-installer extension available [--extension-repository <path>]` = list extensions available for installation
+  * `fuseml-installer extension add <name> [--extension-repository <path>]` - install selected extension
+  * `fuseml-installer extension remove <name> [--extension-repository <path>]` - ubinstalls selected extension
+
```
Comment (jsuchome) [inline docs/blueprints/001-installation-of-extensions.md:49]: Suggestion from @alfsuse :

3. Add information about extensions into the config file

File: docs/blueprints/001-installation-of-extensions.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+```yaml
+name: knative 
+description: |
+  Kubernetes-based platform to deploy and manage modern serverless workloads.
+namespace: knative-serving
+type: script
+install: https://github.com/fuseml/extensions/installation/knative/install.sh
+uninstall: https://github.com/fuseml/extensions/installation/knative/uninstall.sh
+```
```
Comment (jsuchome) [inline docs/blueprints/001-installation-of-extensions.md:100]: TODO we need to solve the case of the combinations, e.g. installing helm chart _and_ some manifest...

**Review by flaviodsr — CHANGES_REQUESTED** [review]
LGTM, there are only two things that I am not sure:

- I think that `installation` is in fact the FuseML extension, so I wonder if it would be better to have everything related to the extension (helm chart, docker image) inside the same directory and have a directory structure like that instead:
```bash
extensions
├── knative
│   ├── description.yaml
│   ├── install.sh
│   └── uninstall.sh
└── mlflow
    ├── description.yaml
    ├── image
    └── mlflow-helm
```
or maybe move the helm charts/images to another repository.

- Regarding supporting install/uninstall through shell scripts might be a bit dangerous, anything can be included in the script and this can include new dependencies. Besides, each system has its own version of bash/sh/... and making sure that the script is compatible with all those system might not be so easy.

**Comment by jsuchome** [comment]
> * so I wonder if it would be better to have everything related to the extension (helm chart, docker image) inside the same directory and have a directory structure like that instead

Hm, interesting, that could be done and indeed the directory structure would be simpler.

However with my proposal, you have a cleaner, somewhat more independent view over different components for example when someone wants to use _just_ the helm chart, without the knowledge about the extensions and how they are used by fuseml...

Also, about the images - aren't we expecting, like _tons_ of images for each extension, based on different configurations, dependencies, etc.? In other words, do you expect to have just one (or few) image(s) per extension?

**Comment by jsuchome** [comment]
> * Regarding supporting install/uninstall through shell scripts might be a bit dangerous, anything can be included in the script and this can include new dependencies.

Of course, but do you have better idea?
Not mentioning that "anything can be included" applies to helm charts and manifest too

File: docs/blueprints/001-installation-of-extensions.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      - description.yaml
+      - install.sh
+      - uninstall.sh
+```
+
+Possible extension types are
+
+* `helm` - helm chart. For helm chart type, no extra install/uninstall commands are necessary
+* `manifest` - Kubernetes manifest, to be installaded using `kubectl`. No extra install/uninstall commands are necessary. All information (including namespace for example) are expected to be present in the manifest file.
+* `script` - shell script. Specific `install` and `uninstall` actions need to be provided by way of referencing specific scripts.
```
Comment (jsuchome) [inline docs/blueprints/001-installation-of-extensions.md:76]: `script` here looks scary, right? But currently we can't get past it, see e.g. how we install knative https://github.com/fuseml/fuseml/blob/main/scripts/knative-install.sh

File: docs/blueprints/001-installation-of-extensions.md (hunk 2 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:100]: See e.g. current seldon-core installation https://github.com/fuseml/fuseml/blob/main/scripts/seldon-operator-install.sh

File: docs/blueprints/001-installation-of-extensions.md (hunk 3 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:76]: We  might agree to have `script` typo for now (so we have working solutions for extension installation soon) and aim for replacing existing scripts with manifests or helm charts.... 

We might actually create our own helm charts, when they are missing.

File: docs/blueprints/001-installation-of-extensions.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ - installation
+    - mlflow
+      - description.yaml
+    - knative
+      - description.yaml
+      - install.sh
+      - uninstall.sh
+```
+
+Possible extension types are
```
Comment (stefannica) [inline docs/blueprints/001-installation-of-extensions.md:72]: I wouldn't call this an extension type, but the type of installation for an extension. You should reserve the name extension type to be used for things like prediction platforms, model stores, etc. But not my point here...

My advice would be to try and see if this maps well to the use-cases we already have: KNative, KFServing and maybe even Seldon Core. I think often you need to combine several installation types and run them in a sort of check-list manner: first, you install a helm chart, then you create some resources via manifests, then maybe you run some command or script. So instead having one installation type here, maybe it would be more flexible to have an ordered list of installation steps.

Of course, you might say that anything that's more complex than a helm chart or a single k8s manifest should be implemented as a shell script.

File: docs/blueprints/001-installation-of-extensions.md (hunk 3 above)
Comment (stefannica) [reply docs/blueprints/001-installation-of-extensions.md:76]: I suggested a hybrid approach in another comment: using a list of installation steps provides us with a smooth transition from one `script` step to several `helm` and `manifest` steps.

File: docs/blueprints/001-installation-of-extensions.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+* `helm` - helm chart. For helm chart type, no extra install/uninstall commands are necessary
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:74]: yeah, `values.yaml` is good point

File: docs/blueprints/001-installation-of-extensions.md (hunk 3 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:76]: Yes, this would definitely help e.g. with seldon-core installation https://github.com/fuseml/fuseml/blob/main/scripts/seldon-operator-install.sh

However not necessary with the others, see e.g. the script for istio installation https://github.com/fuseml/fuseml/blob/main/scripts/istio-minimal-install.sh
(and no, I do not plan to move istio to the extension, just pointing out the example where script seems to be necessary for the time being)

File: docs/blueprints/001-installation-of-extensions.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:76]: I'm not sure I understand your namespace argument: the example yaml file contains extra `namespace` entry...

File: docs/blueprints/001-installation-of-extensions.md (hunk 3 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:78]: Updated with latest commit

File: docs/blueprints/001-installation-of-extensions.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply docs/blueprints/001-installation-of-extensions.md:76]: it does. please ignore that part

File: docs/blueprints/001-installation-of-extensions.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+name: mlflow 
+description: |
+  MLFlow is an open source platform specialized in tracking ML experiments, and packaging and deploying ML models.
+namespace: 
+install:
+  - type: helm
+    url: https://github.com/fuseml/extensions/charts/mlflow
+    values: https://github.com/fuseml/extensions/installation/mlflow/values.yaml
+  - type: manifest
+    url: https://github.com/fuseml/extensions/installation/mlflow/post-install.yaml
```
Comment (stefannica) [inline docs/blueprints/001-installation-of-extensions.md:95]: There seems to be too much duplication needed concerning URLs, especially if you have everything in the same location as the `description.yaml` file. Could the `url` field point to a local path/file that is relative to the `description.yaml` file (or can be absolute, whatever) ? if yes, then please add an example to cover that scenario, something like this:

```
install:
  - type: helm
    url: https://github.com/fuseml/extensions/charts/mlflow
    values: https://github.com/fuseml/extensions/installation/mlflow/values.yaml
  - type: manifest
    url: post-install.yaml
```

Of course, now calling it `url` doesn't work anymore, you'll need another name for it.

File: docs/blueprints/001-installation-of-extensions.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+`script` type seems like a least secure one, and we should aim for replacing it with the other types in the future.
+
+
+#### Examples of extension files
+
+```yaml
+name: mlflow 
+description: |
+  MLFlow is an open source platform specialized in tracking ML experiments, and packaging and deploying ML models.
+namespace:
```
Comment (stefannica) [inline docs/blueprints/001-installation-of-extensions.md:89]: namespace should probably _also_ be a step option, in case you want to install something in another namespace

File: docs/blueprints/001-installation-of-extensions.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+description of operations that could be done with the extension.
+
+Each extension has its own subdirectory that matches the extension name. Mandatory component under each extension subdirectory is the
+description file, in `yaml` format.
+
+Directory structure:
+
+```
+extensions
+ - installation
```
Comment (stefannica) [inline docs/blueprints/001-installation-of-extensions.md:64]: I'd just call it `installer` for now

File: docs/blueprints/001-installation-of-extensions.md (hunk 7 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:89]: yeah, I guess, it does not hurt

**Comment by stefannica** [comment]
> > * so I wonder if it would be better to have everything related to the extension (helm chart, docker image) inside the same directory and have a directory structure like that instead
> 
> Hm, interesting, that could be done and indeed the directory structure would be simpler. I also do not like that `installation` part of the path...
> 
> However with my proposal, you have a cleaner, somewhat more independent view over different components for example when someone wants to use _just_ the helm chart, without the knowledge about the extensions and how they are used by fuseml...
> 
> Also, about the images - aren't we expecting, like _tons_ of images for each extension, based on different configurations, dependencies, etc.? In other words, do you expect to have just one (or few) image(s) per extension?
> 
> > or maybe move the helm charts/images to another repository.
> 
> Yeah, maybe this would be better.

My 2 cents: let's keep these new installer extension files under a separate root folder (suggested simply calling it `installer` for now). When we add more than installation instructions to the extension, we can think about changing the structure of the extensions repo to serve as a sort of "catalog" and group more than just installation artifacts under each sub-directory.

File: docs/blueprints/001-installation-of-extensions.md (hunk 8 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:64]: ok

File: docs/blueprints/001-installation-of-extensions.md (hunk 6 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:95]: `location`

File: docs/blueprints/001-installation-of-extensions.md (hunk 7 above)
Comment (jsuchome) [reply docs/blueprints/001-installation-of-extensions.md:89]: Let's put namespace as optional into the step part, and if it is missing there, the global one will be used. If global one is not specified, it's up to installer.

**Comment by flaviodsr** [comment]
> > * Regarding supporting install/uninstall through shell scripts might be a bit dangerous, anything can be included in the script and this can include new dependencies.
> 
> Of course, but do you have better idea?
> Not mentioning that "anything can be included" applies to helm charts and manifest too

By anything I mean not any kubernetes resource but any call to a *hidden malicious* binary or call to a binary that is not supported by the system running the installer.

Anyways, I think that this is starting point and then we can improve on it later.

**Review by flaviodsr — APPROVED** [review]

File: paas/install_client.go [hunk 9]
```diff
@@ -123,23 +122,100 @@ func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.Installa
...
+	}
+
+	for _, name := range extensions {
+		c.ui.Note().Msg(fmt.Sprintf("Installing extension '%s'...", name))
+
+		extension := deployments.NewExtension(name, extensionRepo.Value.(string))
+
+		err := extension.LoadDescription()
+		if err != nil {
+			return errors.New(fmt.Sprintf("Failed to load description of extension %s: %s", name, err.Error()))
```
Comment (flaviodsr) [inline paas/install_client.go:160]: I think description can be optional, unless there is something depending on it?

File: paas/install_client.go [hunk 10]
```diff
@@ -123,23 +122,100 @@ func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.Installa
...
+
+		err = extension.Install(c.kubeClient, c.ui, options)
+		if err != nil {
+			return errors.New(fmt.Sprintf("Failed to install extension %s: %s", name, err.Error()))
+		}
+	}
+	return nil
+}
+
+// InstallExtensions uninstalls given ML extensions
```
Comment (flaviodsr) [inline paas/install_client.go:171]: typo `InstallExntensions`

File: paas/install_client.go [hunk 11]
```diff
@@ -123,23 +122,100 @@ func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.Installa
...
+func (c *InstallClient) UnInstallExtensions(extensions []string, options *kubernetes.InstallationOptions) error {
```
Comment (flaviodsr) [inline paas/install_client.go:172]: typo: `UnInstall`

File: paas/install_client.go [hunk 12]
```diff
@@ -123,23 +122,100 @@ func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.Installa
...
+	}
+	for _, name := range extensions {
+		c.ui.Note().Msg(fmt.Sprintf("Removing extension '%s'...", name))
+
+		repository := config.DefaultExtensionsLocation()
+		extension := deployments.NewExtension(name, repository)
+
+		err := extension.LoadDescription()
+		if err != nil {
+			return errors.New(fmt.Sprintf("Failed to load description of extension %s: %s", name, err.Error()))
```
Comment (flaviodsr) [inline paas/install_client.go:185]: Same here, in this case I really think there is no neeed to check for description when uninstalling.

File: paas/install_client.go [hunk 13]
```diff
@@ -123,23 +122,100 @@ func (c *InstallClient) Install(cmd *cobra.Command, options *kubernetes.Installa
...
+			return errors.New(fmt.Sprintf("Failed to uninstall extension %s: %s", name, err.Error()))
+		}
+	}
+
+	return nil
+}
+
 // Uninstall removes fuseml from the cluster.
-func (c *InstallClient) Uninstall(cmd *cobra.Command) error {
+func (c *InstallClient) UnInstall(cmd *cobra.Command, options *kubernetes.InstallationOptions) error {
```
Comment (flaviodsr) [inline paas/install_client.go:198]: typo: `UnInstall`, is this intended? why not just `Uninstall`?

File: deployments/extensions.go [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+	Namespace string
+}
+
+type istioGateway struct {
+	Name        string
+	Port        int
+	ServiceHost string
+}
+
+type extensionDesc struct {
```
Comment (flaviodsr) [inline deployments/extensions.go:37]: I think `extensionDesc` is a bit confusing, I would like to suggest `extensionDefinition`.

File: deployments/extensions.go [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+
+// LoadDescription finds the description file of the extension and loads it into the struct
+func (e *Extension) LoadDescription() error {
+
+	u, err := url.Parse(e.Repository)
+	if err != nil {
+		return err
+	}
+
+	tmpDir, err := ioutil.TempDir("", "fuseml-extension")
```
Comment (flaviodsr) [inline deployments/extensions.go:70]: suggestion: a const variable for `fuseml-extension`

File: deployments/extensions.go [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+		descFilePath = filepath.Join(e.Repository, e.Name, defaultDescriptionFileName)
+	}
+
+	// parse and load descriptin file into Extension struct
+	data, err := os.ReadFile(descFilePath)
+	if err != nil {
+		return errors.Wrap(err, "failed to read description file")
+	}
+
+	err = yaml.Unmarshal(data, &e.desc)
```
Comment (flaviodsr) [inline deployments/extensions.go:104]: could you give a try to this yaml package: `github.com/ghodss/yaml` ?


This is what we are using in fuseml/fuseml-core, after we switched from gopkg.in/yaml.v2 due to its issues with capital characters. I am pretty sure you will not have the same problem here, however if one day we move fuseml-core to fuseml at least we would have been using the same yaml library.

File: deployments/extensions.go [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+			return "", err
+		}
+		return filepath.Join(tmpDir, name), nil
+	}
+	// 4. relative path to extension local path
+	return filepath.Join(e.Repository, e.Name, filePath), nil
+}
+
+func (e *Extension) executeScript(path string) error {
+	tmpDir, err := ioutil.TempDir("", "fuseml-extension-chart")
```
Comment (flaviodsr) [inline deployments/extensions.go:148]: suggestion: const variable for `fuseml-extension-chart`

File: deployments/extensions.go [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+
+	if out, err := helpers.RunProc(fullCmd, tmpDir, e.Debug); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed running script: %s\n", out))
+	}
+
+	return nil
+}
+
+func (e *Extension) installManifest(path, ns string) error {
+	tmpDir, err := ioutil.TempDir("", "fuseml-extension-chart")
```
Comment (flaviodsr) [inline deployments/extensions.go:167]: here you are using the same temp directory name as the chart, if this is not a typo maybe remote the `-chart`.

File: deployments/extensions.go [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+
+	out, err := helpers.Kubectl(fmt.Sprintf("apply --filename %s --namespace %s", manifestLocalPath, ns))
+
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("kubectl apply failed:\n%s", out))
+	}
+	return nil
+}
+
+func (e *Extension) unInstallManifest(path, ns string) error {
```
Comment (flaviodsr) [inline deployments/extensions.go:186]: typo: `unInstallManifest`

File: deployments/extensions.go [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+	tmpDir, err := ioutil.TempDir("", "fuseml-extension-chart")
```
Comment (flaviodsr) [inline deployments/extensions.go:187]: using the same folder name again.

File: deployments/extensions.go [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+	out, err := helpers.Kubectl(fmt.Sprintf("delete --filename %s --namespace %s", manifestLocalPath, ns))
+
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("kubectl delete failed:\n%s", out))
+	}
+	return nil
+}
+
+// TODO move under helpers
+func (e *Extension) installHelmPackage(name, chartPath, ns, valuesPath string) error {
```
Comment (flaviodsr) [inline deployments/extensions.go:207]: I would like to suggest naming it `installHelmChart`

File: deployments/extensions.go [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+		return err
+	}
+	if out, err := helpers.RunProc(helmCmd, currentdir, e.Debug); err != nil {
+		return errors.New(fmt.Sprintf("Failed installing %s chart: %s", name, out))
+	}
+
+	return nil
+}
+
+func (e *Extension) unInstallHelmPackage(ui *ui.UI, name, ns string) error {
```
Comment (flaviodsr) [inline deployments/extensions.go:242]: same here, for using `HelmChart` instead of `HelmPackage`

File: deployments/extensions.go [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+		namespace = defaultNamespace
+	}
+	// based on installation type (script/helm/manifest), proceed with uninstallation of each install step
+	for _, step := range e.desc.Uninstall {
+
+		ns := step.Namespace
+		if ns == "" {
+			ns = namespace
+		}
+		switch step.Type {
```
Comment (flaviodsr) [inline deployments/extensions.go:291]: maybe consider adding a default, with some error for a unknown step.

File: deployments/extensions.go [hunk 24]
```diff
@@ -0,0 +1,405 @@
...
+		namespace = defaultNamespace
+	}
+	// based on installation type (script/helm/manifest), proceed with execution of each install step
+	for _, step := range e.desc.Install {
+		ns := step.Namespace
+		if ns == "" {
+			ns = namespace
+		}
+
+		switch step.Type {
```
Comment (flaviodsr) [inline deployments/extensions.go:339]: same here: maybe consider adding a default, with some error for a unknown step.

File: deployments/extensions.go [hunk 25]
```diff
@@ -0,0 +1,405 @@
...
+		err := c.LabelNamespace(
+			e.desc.Namespace,
+			kubernetes.FusemlDeploymentLabelKey,
+			kubernetes.FusemlDeploymentLabelValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO wait for some pod to exist/run? Extra option in the description file
```
Comment (flaviodsr) [inline deployments/extensions.go:377]: why need to wait here?

File: cmd/internal/client/uninstall.go [hunk 26]
```diff
@@ -1,11 +1,29 @@
 package client
 
 import (
+	"github.com/fuseml/fuseml/cli/kubernetes"
 	"github.com/fuseml/fuseml/cli/paas"
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 )
 
+var unInstallOptions = kubernetes.InstallationOptions{
```
Comment (flaviodsr) [inline cmd/internal/client/uninstall.go:10]: `unInstall` again, is it intended?

**Review by flaviodsr — CHANGES_REQUESTED** [review]
Looking good, just some questions/suggestions.

File: Makefile [hunk 27]
```diff
@@ -175,6 +175,9 @@ mlflow-e2e:
 fuseml-install:
 	@./dist/fuseml-installer install
 
+fuseml-install-with-mlflow:
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply Makefile:178]: I don't know, I think original "plain" `fuseml-install:` could also have some usage

File: paas/install_client.go (hunk 9 above)
Comment (jsuchome) [reply paas/install_client.go:160]: The description is not mean the value of description tag, but the description file, `description.yaml` which is crucial ... let me rephrase the message

File: paas/install_client.go (hunk 12 above)
Comment (jsuchome) [reply paas/install_client.go:185]: It's about loading the file same as above. Commit updated to reflect that

File: deployments/extensions.go (hunk 23 above)
Comment (jsuchome) [reply deployments/extensions.go:291]: good point

File: cmd/internal/client/uninstall.go (hunk 26 above)
Comment (jsuchome) [reply cmd/internal/client/uninstall.go:10]: This is for uninstall command, what's wrong?

File: deployments/extensions.go (hunk 14 above)
Comment (jsuchome) [reply deployments/extensions.go:37]: fine

Comment (jsuchome) [reply deployments/extensions.go:37]: although .... it's really called _description_ already in the specs ...

File: deployments/extensions.go (hunk 16 above)
Comment (jsuchome) [reply deployments/extensions.go:104]: I'll try

File: deployments/extensions.go (hunk 18 above)
Comment (jsuchome) [reply deployments/extensions.go:167]: good catch (not that it matters in reality, but  you are right it's not consistent)

File: deployments/extensions.go (hunk 21 above)
Comment (jsuchome) [reply deployments/extensions.go:207]: OK

File: deployments/extensions.go (hunk 22 above)
Comment (jsuchome) [reply deployments/extensions.go:242]: will do

File: deployments/extensions.go (hunk 19 above)
Comment (jsuchome) [reply deployments/extensions.go:186]: What's wrong? You think it should be `uninstall`?

File: deployments/extensions.go (hunk 25 above)
Comment (jsuchome) [reply deployments/extensions.go:377]: We have that kind of code when deploying some other components, take a look at `core.go` for example. It should be fine with helm chart (because I added `--wait` option), but I'm not sure about manifests.

File: paas/install_client.go (hunk 13 above)
Comment (jsuchome) [reply paas/install_client.go:198]: It was intended, do you think it is wrong?

Comment (flaviodsr) [reply paas/install_client.go:198]: As uninstall is a single word I don't think there is a need for the capital `I`

File: deployments/extensions.go (hunk 14 above)
Comment (flaviodsr) [reply deployments/extensions.go:37]: I guess that's the good thing about keeping the spec and the implementation on the same PR as there are some things that we can only see on the code and make changes accordingly.

In any case, feel free to ignore my suggestion if you think that `extensionDesc` fits better.

File: deployments/extensions.go (hunk 19 above)
Comment (flaviodsr) [reply deployments/extensions.go:186]: yes

File: deployments/extensions.go (hunk 25 above)
Comment (flaviodsr) [reply deployments/extensions.go:377]: ok, now I see it, thanks.

For manifests it is indeed difficult to check for the state when we do not know what kind of workloads are created. Maybe the easier way is just to wait for pods to be in the `Ready` state?

Comment (jsuchome) [reply deployments/extensions.go:377]: Yeah, but which pods?
We can either introduce it as a part of installation step description - just list of pods names to wait for.
It is kind of ugly (from user's POV), but easy to implement.

Or we'd need to actually understand the manifest - parse it for pod definitions. This seems ugly from coding POV.

Or we could do something in between, e.g. _wait for all pods in given namespace_ ... but that could be problematic because such namespace could potentially contain some unrelated deployments...

Comment (flaviodsr) [reply deployments/extensions.go:377]: I think we can start with all pods.

Another option is using kustomize to insert specific labels and wait for the pods with that label. But I am not sure if we want to go that way in this PR.

Something to think about, what if we only perform installation from manifests (considering that manifests can be generated from helm/istioctl/(is there something else?) and then we insert the custom labels on those manifests using kustomize. I think this would give an easy way to control the resources created by the extensions.

Comment (jsuchome) [reply deployments/extensions.go:377]: Labels could be good also for listing the extensions. However we would need them for all extension types, not just manifest...

File: cmd/internal/client/uninstall.go (hunk 26 above)
Comment (jsuchome) [reply cmd/internal/client/uninstall.go:10]: ok, you want to have `uninstallOptions`

File: Makefile (hunk 27 above)
[the remark this answers is not in the document]
Comment (jsuchome) [reply Makefile:178]: I kept `fuseml-install-with-mlflow` and fixed `test-mlflow-e2e` target

File: paas/install_client.go (hunk 13 above)
Comment (jsuchome) [reply paas/install_client.go:198]: OK, I was not sure what's the best way of capitalization here, but yep, I will change it

File: deployments/extensions.go (hunk 14 above)
Comment (jsuchome) [reply deployments/extensions.go:37]: Definition could be also used (not just function and scruct names, but also `definition.yaml` file), but I kind of like "description" more...

File: deployments/extensions.go (hunk 25 above)
Comment (jsuchome) [reply deployments/extensions.go:382]: I would rather solve this in separate PR

**Review by flaviodsr — APPROVED** [review]
LGTM
